Modernize Indigo

PepperCode1 opened the pull request:
- Fix `QuadView#copyTo` not copying enough data
  - The specification (Javadoc) says that this method copies everything except for the material. In reality, it was copying everything except the header bits, which include the cull face, light face, normal set flags, geometry flags, and material, and then marking the target quad's geometry as valid. This did not make sense as the quad data which includes vertex positions was replaced but the header bits which include the geometry were not.
- Refactor `ItemRenderContext`
  - Fix enchantment glint not rendering on item models rendered with this context
  - Make vertex consumer calculation mirror vanilla
  - Improve vertex consumer caching
  - Restructure quad rendering to be more efficient and consistent with block rendering
- Remove unnecessary code including `CompatibilityHelper`
- Add `@Unique` annotations to fields added by Indigo
- Fix typos
- Organize imports

I have tested these changes with Continuity, a mod that heavily utilizes the renderer API, and did not notice any new issues.

PepperCode1 committed:
Modernize Indigo

- Restructure ItemRenderContext to be more efficient and consistent with block rendering
- Remove unnecessary code including CompatibilityHelper
- Add Unique annotations to fields added by Indigo
- Fix typos
- Organize imports

PepperCode1 committed:
Remove unused import

PepperCode1 committed:
Fix bugs

- Fix enchantment glint not rendering on item models
- Fix QuadView#copyTo not copying enough data
- Make ItemRenderContext vertex consumer calculation mirror vanilla and be more efficient

PepperCode1 committed:
Clear the target quad's material

comp500 commented on fabric-renderer-indigo/src/main/java/net/fabricmc/fabric/impl/client/indigo/renderer/mesh/QuadViewImpl.java:
```diff
@@ -163,7 +163,8 @@ public void copyTo(MutableQuadView target) {
 		// copy everything except the material
-		System.arraycopy(data, baseIndex + 1, quad.data, quad.baseIndex + 1, EncodingFormat.TOTAL_STRIDE - 1);
+		System.arraycopy(data, baseIndex, quad.data, quad.baseIndex, EncodingFormat.TOTAL_STRIDE);
+		quad.data[quad.baseIndex + HEADER_BITS] = EncodingFormat.clearMaterial(quad.data[quad.baseIndex + HEADER_BITS]);
```
Should this preserve the existing material stored in the `MutableQuadViewImpl` rather than clearing it?

PepperCode1 replied:
I suppose so.

comp500 reviewed (changes requested):
Looks good, though I wonder if the `@Unique` mixin fields should be renamed to have a consistent naming scheme, since some have a `fabric_` prefix and some don't (I'm not sure what FAPI convention is). Similarly for final fields, should those be capitalised?

PepperCode1 commented:
> Looks good, though I wonder if the `@Unique` mixin fields should be renamed to have a consistent naming scheme, since some have a `fabric_` prefix and some don't (I'm not sure what FAPI convention is). Similarly for final fields, should those be capitalised?

This can be done, but the convention is that only static final fields are capitalized, which has been done.

PepperCode1 committed:
Tweaks

- Retain material during copyTo instead of clearing it
- Standardize mixins

PepperCode1 committed:
Direct return

- Return from ItemRenderContext#quadVertexConsumer instead of assigning value to variable
